lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy/Gen.hs

```diff
@@ -0,0 +1,58 @@
+module Cardano.Wallet.Primitive.Types.TokenPolicy.Gen
+    ( genTokenNameSmallRange
+    , genTokenPolicyIdSmallRange
+    , shrinkTokenNameSmallRange
+    , shrinkTokenPolicyIdSmallRange
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types.TokenPolicy
+    ( TokenName, TokenPolicyId )
+import Cardano.Wallet.Unsafe
+    ( unsafeFromHex )
+import Test.QuickCheck
+    ( Gen, elements )
+
+import qualified Cardano.Wallet.Primitive.Types.TokenPolicy as TP
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text as T
+
+--------------------------------------------------------------------------------
+-- Token names chosen from a small range (to allow collisions)
+--------------------------------------------------------------------------------
+
+genTokenNameSmallRange :: Gen TokenName
+genTokenNameSmallRange = elements tokenNames
+
+shrinkTokenNameSmallRange :: TokenName -> [TokenName]
+shrinkTokenNameSmallRange name = filter (< name) tokenNames
+
+tokenNames :: [TokenName]
+tokenNames = TP.mkTokenName . ("Token" `T.snoc`) <$> ['A' .. 'D']
+
+--------------------------------------------------------------------------------
+-- Token policy identifiers chosen from a small range (to allow collisions)
+--------------------------------------------------------------------------------
+
+genTokenPolicyIdSmallRange :: Gen TokenPolicyId
+genTokenPolicyIdSmallRange = elements tokenPolicies
+
+shrinkTokenPolicyIdSmallRange :: TokenPolicyId -> [TokenPolicyId]
+shrinkTokenPolicyIdSmallRange policy = filter (< policy) tokenPolicies
+
+tokenPolicies :: [TokenPolicyId]
+tokenPolicies = mkTokenPolicyId <$> ['A' .. 'D']
+
+--------------------------------------------------------------------------------
+-- Utilities
+--------------------------------------------------------------------------------
+
+mkTokenPolicyId :: Char -> TokenPolicyId
+mkTokenPolicyId
+    = TP.mkTokenPolicyId
+    . unsafeFromHex
+    . B8.replicate tokenPolicyIdHexStringLength
```

Comment:
It might be nicer for QC counterexamples if these hashes are just single characters in a..z, or something like that?

Reply:
@rvl I did consider using just a single character, but the `FromText` instance for `TokenPolicyId` is based on the `Hash` type, which expects to find a byte string of the correct length. If we change the existing generator to use just one character, then the textual serialization tests will fail.

The generator and shrinker currently produce values like this:
* `aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa`
* `bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb`

But I can see it would be nicer to have:
* `PolicyA`
* `PolicyB`
  ... and so on.

Of course, we **_could_** split this generator up into two different generators:

1. Generate valid-length hashes for serialization tests.
2. Generate short hashes for all other tests.

Or, we could use `ShowFmt` in the tests, and just display the first `n` characters in the event of a test failure?

Reply:
OK, that's too bad then. I have found that full-length hashes make counterexamples more difficult to understand.

Reply:
> OK, that's too bad then. I have found that full-length hashes make counterexamples more difficult to understand.

I agree with this.

Though perhaps we can solve this by adding a `ShowFmt` to the tests? We could easily make the `Buildable` instances just display the first `n` bytes of each hash (or perhaps the prefix and suffix: `aaaa...aaaa`).